### PR TITLE
Feature/search

### PR DESCRIPTION
**Pull request by pavelkondratyev**
Created search route and logic

### CODE REVIEW

File: middlewares/search.middleware.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * @function parseQuery
+ * @param {JSON} req
+ * @param {JSON} res
+ * @param {JSON} next
+ * @description parses the json in the parameter
+ */
+function parseQuery(req, res, next){
+    let query = req.query.q;
```
Comment (pierreTklein) [inline middlewares/search.middleware.js]: You might want to grab the query from req.body-- middleware.parseBody will put the validated query 'q' into there. Let's chat about this

File: middlewares/search.middleware.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+    req.q = JSON.parse(query);
+    req.model = req.query.model;
```
Comment (pierreTklein) [inline middlewares/search.middleware.js]: Same here

File: middlewares/validators/validator.helper.js [hunk 3]
```diff
@@ -336,8 +339,60 @@ function jwtValidator (fieldLocation, fieldname, jwtSecret, optional = true) {
...
+    const search = setProperValidationChainBuilder(fieldLocation, fieldname, "Invalid search query");
+    return search.exists().withMessage("Search query must be provided")
+        .custom((value) => {
+            //value is a serialized JSON
+            value = JSON.parse(value);
+            let modelString = query("model", "Corresponding model not found");
+            let model;
+            //Supported models for searching
+            if(modelString.equals('hacker')){
+                model = Models.Hacker;
```
Comment (pierreTklein) [inline middlewares/validators/validator.helper.js]: So this search is just for Hackers, right

File: middlewares/validators/validator.helper.js [hunk 4]
```diff
@@ -336,8 +339,60 @@ function jwtValidator (fieldLocation, fieldname, jwtSecret, optional = true) {
...
+                //Checks that each clause has Param, Value, and Operation
+                var clause = value[q];
+                if(!(clause.hasOwnProperty("param") || clause.hasOwnProperty("value") || clause.hasOwnProperty("operation"))){
+                    return false;
+                }
+                var schemaPath = model.searchableField(clause.param);
+                if(!schemaPath) return false;
+                //Validates that operation corresponding to each clause is valid
+                switch (schemaPath) {
+                    case 'String':
```
Comment (pierreTklein) [inline middlewares/validators/validator.helper.js]: Are they only allowed to search by `String`, `Number`, and `Boolean`?

File: routes/api/search.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+const Middleware = {
+    Validator: {
+        Search: require("../../middlewares/validators/search.validator")
+    },
+    parseBody: require("../../middlewares/parse-body.middleware"),
+    Search: require('../../middlewares/search.middleware')
+};
+
+module.exports = {
+    /**
```
Comment (pierreTklein) [inline routes/api/search.js]: Small knit-pick: Move this comment to right above 
```javascript
searchRouter.route("/").get(
```

File: routes/api/search.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * @apiName search
+     * @apiGroup Search
+     * @apiVersion 0.0.8
+     *
+     * @apiParam (query) {String} model the model to be searched
+     * @apiParam (query) {Array} q the query to be executed
+     *
+     * @apiSuccess {String} message Success message
+     * @apiSuccess {Object} data Results
+     * @apiSuccessExample {object} Success-Response:
```
Comment (pierreTklein) [inline routes/api/search.js]: You might also want to provide another example for when the message returns "no items found" or whatever

File: services/search.service.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+                query.where(param).regex(val);
+                break;
+            case "elemMatch":
+                query.where(param).elemMatch(val);
+                break;
+        }
+    }
+    return await query.lean()
+        .exec()
+        .catch(
```
Comment (pierreTklein) [inline services/search.service.js]: Not sure what happens when you call `await` on a Promise that has `.catch` already bound. You might just want to do 
```javascript
return query.lean().exec()
```
and make this function not asynchronous. As long as the async middleware is wrapped by the util.asyncMiddleware function, rejected promises will be handled.

File: tests/search.service.spec.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        operation: "equals",
+        value: "Female"
+    }
+]
+
+const badQuery =[
+    {
+        param: "password",
+        operation: "equals",
+        value: "passowrd"
```
Comment (pierreTklein) [inline tests/search.service.spec.js]: lmao love this, maybe `pA$$w0rD`? jk

File: tests/search.service.spec.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    it("Should return an error as hackers don't have password stored", function(done) {
+        chai.request(server.app)
+            .get("/api/search/")
+            .query({q: JSON.stringify(badQuery), model: "hacker"})
+            .end(function (err, res) {
+                res.should.have.status(422);
+                done();
+            });
+    })
+    it("Should return an error as staff aren't searchable", function(done) {
```
Comment (pierreTklein) [inline tests/search.service.spec.js]: In the future, Admins should have the ability to search staff, etc

File: middlewares/search.middleware.js (hunk 2 above)
Comment (pavelkondratyev) [reply middlewares/search.middleware.js]: Yeah I like it

File: tests/search.service.spec.js (hunk 9 above)
Comment (pavelkondratyev) [reply tests/search.service.spec.js]: Oh yeah just wanted to give an example that diff models aren't searchable yet

File: middlewares/validators/validator.helper.js (hunk 3 above)
Comment (pavelkondratyev) [reply middlewares/validators/validator.helper.js]: Yeah but it's very easy, can add whichever ones

File: tests/search.service.spec.js (hunk 9 above)
Comment (pierreTklein) [reply tests/search.service.spec.js]: Sounds good!

File: middlewares/search.middleware.js [hunk 10]
```diff
@@ -11,17 +15,27 @@ const Services = {
...
-    delete req.query.q;
+    next();
+}
 
+/**
+ * Middleware that executes the query passed
+ * @param {{body: {model: string, q: Array}}} req
+ * @param {*} res
+ * @param {(err?)=>void} next
+ * @returns {Promise.<void>}
```
Comment (pierreTklein) [inline middlewares/search.middleware.js]: Small: no need for return statement

**Pull request merged**